Convert attrd server-side IPC to new model

clumens opened the pull request.

clumens commented:
Let's just go ahead and call this a draft for the moment.

kgaillot commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -482,6 +459,14 @@ attrd_ipc_dispatch(qb_ipcs_connection_t * c, void *data, size_t size)
+
+        if (client->name == NULL) {
+            const char *value = crm_element_value(xml, F_ORIG);
+            client->name = crm_strdup_printf("%s.%d", value?value:"unknown", client->pid);
```
Can use pcmk__s() here

kgaillot reviewed (commented):
This is a good start. I didn't trace the flags in detail but that's often where problems arise, the flags are often different in places you wouldn't expect them to be.

attrd_broadcast_protocol() and attrd_send_message() now belong in attrd_messages.c

I think we could probably consolidate some functions now. If there's nothing IPC or CPG specific in the handler, it should probably be moved from ipc/corosync.c directly into the handler in messages.c. Some longer functions might still be worth multiple functions. In other words messages.c would handle more of the semantics of messages whereas ipc.c and corosync.c would handle the underlying communication mechanism more generically. But that might not be ideal, just one possibility that comes to mind.

One nitpick, it's not really new-style IPC, more new-style message handling, since they can be IPC or CPG messages

clumens commented on daemons/attrd/attrd_messages.c:
```diff
@@ -0,0 +1,86 @@
+                                pcmk_exec_status_str(request->result.execution_status),
+                                (reason == NULL)? "" : " (",
+                                (reason == NULL)? "" : reason,
+                                (reason == NULL)? "" : ")");
```
Fixed.

clumens commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -205,57 +205,39 @@ attrd_client_peer_remove(pcmk__client_t *client, xmlNode *xml)
-        }
-    }
-    free_xml(reply);
+    crm_log_xml_trace(reply, "Reply");
```
It was actually only being logged for that one message, but I've moved it to attrd_handle_request so it'll get logged for all of them.

clumens commented on daemons/attrd/attrd_utils.c:
```diff
@@ -46,7 +48,12 @@ attrd_broadcast_protocol(void)
+    crm_xml_add(attrd_op, PCMK__XA_ATTR_NODE_NAME, host);
+    crm_xml_add_int(attrd_op, PCMK__XA_ATTR_NODE_ID, attrd_cluster->nodeid);
+
+    attrd_send_message(NULL, attrd_op); /* ends up at attrd_peer_message() */
```
Added a debug message back in specifically for this.

clumens commented on daemons/attrd/attrd_messages.c:
```diff
@@ -0,0 +1,86 @@
+    if (reply != NULL) {
+        pcmk__ipc_send_xml(request->ipc_client, request->ipc_id, reply,
+                           crm_ipc_flags_none);
+        free_xml(reply);
```
Unless I'm missing something, none of the CPG messages on main currently send a response.  But yeah, I can put a condition around this code to only call pcmk__ipc_send_xml for IPC.

clumens commented on daemons/attrd/attrd_utils.c:
```diff
@@ -38,6 +38,8 @@ int minimum_protocol_version = -1;
 void
 attrd_broadcast_protocol(void)
 {
+    char *host = strdup(attrd_cluster->uname);
```
Fixed.

clumens commented:
What might be some good ways to test this set?  I've run it through cts-lab and that's all good so far, but I'm not sure if that is really comprehensive for this case.

kgaillot commented:
> What might be some good ways to test this set? I've run it through cts-lab and that's all good so far, but I'm not sure if that is really comprehensive for this case.

The lab is the most important, and you should be able to test everything else using attrd_updater.

kgaillot commented:
We don't currently have a regression test for pacemaker-attrd the way we do for pacemaker-execd and pacemaker-fenced. It would be nice to add one someday, but the main complication is that it interacts with the CIB, and we wouldn't want a regression test to change the real CIB. We'd have to add a daemon start-up option for using an alternative CIB location. The corosync integration would be another complication, we'd have to temporarily set it up as a single-node cluster, or disable corosync use altogether. I'll create a task ...

clumens commented:
Updated with a couple patches at the end, changed terminology in commit messages to refer to "new message handling", and moved some functions around.  I've run everything through cts-lab and various attrd_updater tests but I still feel like I'm not doing enough.  I'm not quite sure where to go on the function consolidation side of things.

This is probably good enough to no longer mark as a draft, though.

kgaillot commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -153,9 +150,8 @@ attrd_client_clear_failure(pcmk__request_t *request)
     }
 
-    attrd_client_update(xml);
     pcmk__set_result(&request->result, CRM_EX_OK, PCMK_EXEC_DONE, NULL);
```
Do we still need this now that attrd_client_update() does it?

clumens replied:
Nope, definitely not.

kgaillot commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -481,6 +462,14 @@ attrd_ipc_dispatch(qb_ipcs_connection_t * c, void *data, size_t size)
+        if (client->name == NULL) {
+            const char *value = crm_element_value(xml, F_ORIG);
+            client->name = crm_strdup_printf("%s.%d", pcmk__s(value, "unknown"), client->pid);
+        }
```
I suspect we don't need this now that we use (or should use) pcmk__client_name() everywhere

clumens replied:
It looks like the only places left referencing "client->name" are here, various places in mainloop.c that are unrelated, and several spots in the based and execd code which I do not think have been converted over yet.  So yeah this can go.

kgaillot replied:
Yeah only the attrd uses are relevant here

kgaillot commented on daemons/attrd/attrd_messages.c:
```diff
@@ -90,6 +90,25 @@ handle_refresh_request(pcmk__request_t *request)
+         * here to ensure that the client only ever receives one.
+         */
+        attrd_send_ack(request->ipc_client, request->ipc_id, request->flags);
+        return attrd_client_update(request);
```
Another possibility would be to add an ack_needed argument to attrd_client_update(). Not necessary, just thinking out loud

kgaillot commented on daemons/attrd/attrd_utils.c:
```diff
@@ -29,38 +29,7 @@ static bool requesting_shutdown = false;
-    free(host);
-    free_xml(attrd_op);
-}
+extern int minimum_protocol_version;
```
probably worth moving this to the header for future use

kgaillot commented on lib/common/ipc_client.c:
```diff
@@ -1551,3 +1548,28 @@ pcmk__ipc_is_authentic_process_active(const char *name, uid_t refuid,
+        CRM_ASSERT(host != NULL);
+
+        crm_xml_add(request, PCMK__XA_ATTR_NODE_NAME, host);
+        free(host);
```
No need to strdup, crm_xml_add() adds a copy

kgaillot commented on lib/common/ipc_client.c:
```diff
@@ -1551,3 +1548,28 @@ pcmk__ipc_is_authentic_process_active(const char *name, uid_t refuid,
+ * \param[in]     nodeid   The local node's ID (can be 0)
+ */
+void
+pcmk__add_node_to_request(xmlNode *request, const char *node, int nodeid)
```
Nice that this came in so handy.

For the name, I've been using "xe" in any function that manipulates XML elements. Also, this is not particularly IPC- or even request-specific; xml.c, nvpair.c, or a new nodes.c would be better. Maybe pcmk__xe_add_node().

clumens replied:
Renamed and added to nodes.c.

kgaillot commented on tools/attrd_updater.c:
```diff
@@ -255,6 +256,10 @@ main(int argc, char **argv)
+    } else if ((options.command == 'B'|| options.command == 'Y') && options.attr_dampen == NULL) {
+        exit_code = CRM_EX_USAGE;
+        g_set_error(&error, PCMK__EXITC_ERROR, exit_code, "Command requires --delay argument");
+        goto done;
```
I'm a little wary of breaking backward compatility in behavior. Maybe just output a warning. The API will handle a null dampen without a problem.

clumens replied:
Converted this into just printing out a warning.

kgaillot commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -335,6 +335,13 @@ attrd_client_update(pcmk__request_t *request)
 
+    if (a == NULL && pcmk__str_eq(request->op, PCMK__ATTRD_CMD_UPDATE_DELAY, pcmk__str_none)) {
+        pcmk__format_result(&request->result, CRM_EX_ERROR, PCMK_EXEC_ERROR,
+                            "Attribute %s does not exist", attr);
```
CRM_EX_NOSUCH would be better

kgaillot reviewed (commented):
I think it's coming together now

We need to repeat the attrd_updater testing with newer cluster + older remote and vice versa. Testing a mixed-version cluster (older and newer cluster nodes) might also be nice but I don't think anything here could affect that.

clumens commented on daemons/attrd/attrd_messages.c:
```diff
@@ -0,0 +1,89 @@
+            pcmk__ipc_send_xml(request->ipc_client, request->ipc_id, reply,
+                               crm_ipc_flags_none);
+            free_xml(reply);
+        }
```
Logging is a good idea.  I'd say that it should maybe even be an error - a reply from the CPG side here indicates to me that I haven't thought of something.  Either that reply needs to be sent to the client, or it doesn't need to be generated in the first place.

clumens commented on daemons/attrd/attrd_messages.c:
```diff
@@ -0,0 +1,89 @@
+        if (request->ipc_client != NULL) {
+            crm_log_xml_trace(reply, "Reply");
+            pcmk__ipc_send_xml(request->ipc_client, request->ipc_id, reply,
+                               crm_ipc_flags_none);
```
I can't see why I wouldn't have wanted flags here.  Should this be request->flags or request->ipc_flags?

kgaillot replied:
That's tricky. You'll have to compare all the places where this was sent before, and see if they all used the same thing.

clumens replied:
Okay, looks like this should be ipc_flags.  Fixed.

clumens commented:
> We need to repeat the attrd_updater testing with newer cluster + older remote and vice versa. Testing a mixed-version cluster (older and newer cluster nodes) might also be nice but I don't think anything here could affect that.

I tested the following commands on the remote with mismatched versions and log files on both sides look good:

* attrd_updater -R
* attrd_updater -Q --name doesnexist
* attrd_updater -U 123 --name XYZ
* attrd_updater -Q --name XYZ
* attrd_updater -Y --name XYZ -d 1000
* attrd_updater -B 456 -d 999 --name XYZ
* attrd_updater -D --name XYZ

I'm currently running another round of cts-lab tests.  It would be possible for me to test that as mixed-version, too.

kgaillot commented:
> I'm currently running another round of cts-lab tests. It would be possible for me to test that as mixed-version, too.

The lab checks for specific log messages, so it requires all nodes to have the same version. If none of the wording it looks for changed between two versions, mixing them is fine but usually not worth the trouble of figuring that out.

kgaillot commented on tools/attrd_updater.c:
```diff
@@ -255,6 +256,8 @@ main(int argc, char **argv)
         g_set_error(&error, PCMK__EXITC_ERROR, exit_code, "Command requires --name argument");
         goto done;
+    } else if ((options.command == 'B'|| options.command == 'Y') && options.attr_dampen == NULL) {
+        fprintf(stderr, "Warning: '%c' command given without required --delay\n", options.command);
```
Why does attrd_updater still use fprintf()?

clumens replied:
It certainly shouldn't.  I'll fix this usage and then there's two "Could not connect to attrd" strings that I'll fix in a new patch.

clumens replied:
Updated and pushed.

kgaillot reviewed (commented):
looks good

kgaillot commented:
Woohoo!
